doc: add common recommendations and suggestions (#4675)
operator-framework/operator-sdk: website/content/en/docs/best-practices/common-recommendation.md

---
title: "Common recommendations and suggestions"
linkTitle: "Common suggestions"
weight: 2
description: Common recommendations and suggestions to built solutions with Operator SDK
---

## Overview

> Any recommendations or good practices suggested by the Kubernetes community, such as to develop [Operator pattern][operator-pattern] solutions or to use [controller-runtime][controller-runtime], also are good recommendations for those who are looking to build the operator projects with SDK. Also, see [Operator Best Practices][operator-best-practices].

Reviewer: I don't understand this at all. You link a bunch of different pages related to operators,  but don't really mention what THIS page is going to be about. Maybe something like

"This page covers common recommendations when developing operators. They can be helpful when whether you are using operator-sdk directly, developing an application that follows the [operator pattern][operator-pattern], or directly using [controller-runtime][controller-runtime]. Also, please see [Operator Best Practices](operator-best-practices]."

Author: changed as we spoke about tks for help.

## Common Recommendations

### Develop idempotent reconciliation solutions

> See that following the operator's pattern; you will create controllers, from [controller-runtime][controller-runtime], with a reconcile function responsible for synchronizing the resources until a desired state on the cluster. So, it is recommended to develop idempotent solutions for your reconciliations.

Reviewer 2: ```suggestion
See that following the operator's pattern. You will create controllers, from [controller-runtime][controller-runtime], with a reconcile function responsible for synchronizing the resources until there's desired state in the cluster. So, it is recommended to develop idempotent solutions for your reconciliations.
```

> See that following the operator's pattern. You will create controllers, from [controller-runtime][controller-runtime], with a reconcile function responsible for synchronizing the resources until there's desired state in the cluster. So, it is recommended to develop idempotent solutions for your reconciliations.

Reviewer: You don't mention idempotency until the end, i would lead with it. something like:

"When developing operators, it is important for the controllers reconciliation loop to be idempotent. [Controller-runtime's][controller-runtime] Reconcile function is responsible for synchronizing resources until the desired state has been reached, and can be invoked with the object in any possible actual and desired state. If your reconciliation loop is not idempotent, it may be possible for your controller or a watched resource to become "stuck" and require manual intervention.

> ### Understand the API's concept

Reviewer: I think mentioning this is specifically about Kubernetes API would be good. Maybe something like

## Understand Kubernetes APIs

Usually, an operator will consist of an a controller which is responsible for the reconciliation of a [Kubernetes API extension from a Custom Resource Definition.][k8s-crd-doc] Check the Kubebuilder documentation [Groups and Versions and Kinds, oh my!][kb-gkv] to get an idea of how to design better solutions.

Author: solved as we chat bout tks for the help.

> When developing operators, it is important for the controller's reconciliation loop to be idempotent. By following the [Operator pattern][operator-pattern] you will create [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/) which provide a reconcile function responsible for synchronizing resources until the desired state is reached on the cluster.

Reviewer: woops, think we lost a bit in the shuffle here. Would add a line to the end about why reconciliation being idempotent is important. something like

"As a reconciliation loop can begin with the system in any possible state, it is important for your loop to be idempotent, else the controller or a watched resource maybe become stuck and require manual intervention."

Author: I am not sure if I follow your suggestion here.

Reviewer: I'd like to have a line that explains WHY is is important for your loop to be idempotent.

Author: It is important because of `provide a reconcile function responsible for synchronizing resources until the desired state is reached on the cluster.` If you are not looking for the desired state then, it will not be idempotent imo

Author: Many issues scenarios (bad side effects) can be faced if they do not make the impl looking for a final desired state == idempotent since it goes explicitly against its purpose. So, I do not think that we can describe here the possible issues. However, we can put that it will go explicitly against its purpose and the design proposed by controller-runtime. Would that make it better?

Author: I think I could address your concern here. Please, let me know wdyt? 
closing that

> Not necessary but usually, the primary resource managed by these controllers which are reconciled are from your own APIs which [Extend the Kubernetes API with CustomResourceDefinitions][k8s-crd-doc]. So, check  documentation [Groups and Versions and Kinds, oh my!][kb-gkv] to have a better idea and be able to design better solutions.

Reviewer 2: ```suggestion
Usually, the primary resource managed by these controllers which are reconciled is from your APIs, which [Extend the Kubernetes API with CustomResourceDefinitions][k8s-crd-doc]. Check the documentation [Groups and Versions and Kinds, oh my!][kb-gkv] to get a better idea of how to design better solutions.
```

> Building your own operator commonly involves extending the Kubernetes API itself. It is helpful to understand exactly how [Custom Resource Definitions interact with the Kubernetes API][k8s-crd-doc]. Also, the Kubebuilder documentation on [Groups and Versions and Kinds] may be helpful to better understand these concepts as they relate to operators.

Reviewer: think we broke the 2nd link here

### Avoid a design solution where more than one Kind is reconciled by the same controller

> Have many Kinds, for example many CRDs, which are all managed by the same controller such as `install.go` usually goes against the design proposed by [controller-runtime][controller-runtime] beyond might hurt concepts such as; encapsulation, the Single Responsibility principle, and Cohesion. Damaging these concepts could cause unexpected side effects, such as; difficulty extending, to reuse, and low maintainability, only to mention a few.

Reviewer 2: ```suggestion
Have many Kinds, for example, many CRDs, which are all managed by the same controller such as `install.go` usually goes against the design proposed by [controller-runtime][controller-runtime] beyond might hurt concepts such as encapsulation, the Single Responsibility Principle, and Cohesion. Damaging these concepts could cause unexpected side effects, such as difficulty extending, reuse, and low maintainability, only to mention a few.
```

> ### Others common suggestions

Reviewer: should be "Other common suggestions"

> Additionally, we can share:

Reviewer: i dont think we need this line? bullet points are fine

- Provide the images and tags used by the operator solution via environment variables in the `config/manager/manager.yaml`: 

```yaml
...
spec:
  ...
    spec:
      ...
      containers:
      - command:
        - /manager
        ...
        env:
        - name: MY_IMAGE
          value: "quay.io/example.com/image:0.0.1"
```

- Manage your solutions using [Status Conditionals][status-conditionals] 
- Use [finalizers][finalizers] when/if required 
- Cover the project with tests/CI to ensure its quality:
    - For any language-based operator, you can use [Scorecard][scorecard] to implement functional tests
    - For Go-based operators, you can also use [envtest][envtest] to cover the controllers. For further information see [Testing with EnvTest][testing-with-envtest]

>     - For Ansible-based operators, you can also use [Molecule][molecule]. For further information see [Testing with Molecule][molecule-tests]

Reviewer: i would put a blurb about what molecule is. like
"you can also use [Molecule][molecule], an Ansible testing framework."

    - For Helm-based operators, you can also use [Chart tests][helm-chart-tests]

> - Ensure that you checked the [FAQ question Can I customize the projects generated with SDK tool?][faq] and understand the [Project Layout][project-layout] before start to do your customizations as please you on top.

Reviewer 2: ```suggestion
- Ensure that you checked the [FAQ question Can I customize the projects generated with SDK tool?][faq] and understand the [Project Layout][project-layout] before starting to do your customizations as please you on top.
```

- If you will integrate your project with OLM then, also check its [Best Practices][olm-best-practices] section.

>  

Reviewer 2: A teammate has [written an article](https://www.openshift.com/blog/7-best-practices-for-writing-kubernetes-operators-an-sre-perspective) about some best practices for writing an operator (from an SRE perspective). Could be added as a reference in one of the bullet points, or in a separate section for further reading. WDYT?

Author: I think we need to stick with more official links. WDYT about after we get this merged he push a PR with suggested items that would fit here? and/or collab with https://operator-framework.github.io/community-operators/best-practices/ to make that better instead of creating another doc?

See that the idea here is to have what is more specific to sdk and then, we can link that one which is very generic.

Reviewer 2: That sounds like a plan @camilamacedo86. SGTM.
